File: chapt5/chapt5_ws/src/demo_python_tf/demo_python_tf/static_tf_broadcaster.py
import math
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import TransformStamped
from tf2_ros import StaticTransformBroadcaster
from tf_transformations import quaternion_from_euler


class StaticTFBroadcaster(Node):
    def __init__(self):
        super().__init__('static_tf2_broadcaster')
        self.static_broadcaster = StaticTransformBroadcaster(self)
        self.pub_static_tf()

    def pub_static_tf(self):
        # Create a TransformStamped message
        t = TransformStamped()

        # Set the header frame_id and child_frame_id
        t.header.stamp = self.get_clock().now().to_msg()
        t.header.frame_id = 'base_link'
        t.child_frame_id = 'camera_link'

        # Set the translation (x, y, z)
        t.transform.translation.x = 0.5
        t.transform.translation.y = 0.3
        t.transform.translation.z = 0.6

        # Set the rotation (roll, pitch, yaw)
        roll = math.radians(180)
        pitch = 0.0
        yaw = 0.0
        quaternion = quaternion_from_euler(roll, pitch, yaw)
        
        t.transform.rotation.x = quaternion[0]
        t.transform.rotation.y = quaternion[1]
        t.transform.rotation.z = quaternion[2]
        t.transform.rotation.w = quaternion[3]

        # Publish the static transform
        self.static_broadcaster.sendTransform(t)
        self.get_logger().info('Published static transform from world to robot_base')
    
def main(args=None):
    rclpy.init(args=args)
    node = StaticTFBroadcaster()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

File: chapt5/chapt5_ws/src/demo_cpp_tf/src/static_tf_broadcaster.cpp
#include <memory>
#include "rclcpp/rclcpp.hpp"
#include "tf2/LinearMath/Quaternion.h"
#include "tf2_ros/static_transform_broadcaster.h"
#include "geometry_msgs/msg/transform_stamped.hpp"
#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp"


class StaticTfBroadcasterNode : public rclcpp::Node
{
public:
  StaticTfBroadcasterNode()
  : Node("static_tf_broadcaster")
  {
    static_broadcaster_ = std::make_shared<tf2_ros::StaticTransformBroadcaster>(this);
    this->publish_tf();
  }

  void publish_tf()
  {
    geometry_msgs::msg::TransformStamped static_transform_stamped;

    static_transform_stamped.header.stamp = this->get_clock()->now();
    static_transform_stamped.header.frame_id = "map";
    static_transform_stamped.child_frame_id = "target_point";

    static_transform_stamped.transform.translation.x = 5.0;
    static_transform_stamped.transform.translation.y = 3.0;
    static_transform_stamped.transform.translation.z = 0.0;

    tf2::Quaternion quat;
    quat.setRPY(0, 0, 60 * M_PI / 180);  // Roll, Pitch, Yaw
    static_transform_stamped.transform.rotation = tf2::toMsg(quat);

    static_broadcaster_->sendTransform(static_transform_stamped);
    RCLCPP_INFO(this->get_logger(), "Published static transform from 'map' to 'target_point'");
  }
private:
  std::shared_ptr<tf2_ros::StaticTransformBroadcaster> static_broadcaster_;
};  

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<StaticTfBroadcasterNode>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}






